chain-impl-mockchain/src/txbuilder.rs

```diff
@@ -0,0 +1,174 @@
+        match policy {
+            OutputPolicy::Forget => {
+                let tx = TransactionFinalizer(tx::SignedTransaction {
+                    transaction: self.0,
+                    witnesses: vec![],
+                });
+                Ok((Balance::Positive(pos), tx))
+            }
+            // We will try to find the best matching value, for
+            // this reason we will try to reduce the set using
+            // value estimated by the current fee.
+            //
+            // We are searching in a range
+            //   [min_value, max_value)
+            OutputPolicy::One(address) => {
+                // This is simplified version of the algorithm that
+                // works only in case in fee can perfectly estimate
+                // the required cost. We add an additional empty output
+                // hoping that it doesn't change fee count.
+                //
+                // Otherwise better estimation algorithm is needed.
+                let mut tx = self.0.clone();
+                tx.outputs.push(tx::Output {
+                    address: address.clone(),
+                    value: Value(0),
+                });
+                let fee = fee_algorithm
+                    .calculate_for(&tx)
+                    .ok_or(Error::MathErr(ValueError::Overflow))?;
+                match balance(&tx, fee) {
+                    Ok(Balance::Positive(value)) => {
+                        self.0.outputs.push(tx::Output { address, value });
+                        return Ok((Balance::Zero, TransactionFinalizer::new(self.0)));
+                    }
+                    _ => {
+                        return Ok((Balance::Positive(pos), TransactionFinalizer::new(self.0)));
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Amount of the balance in the transaction.
+pub enum Balance {
+    /// Balance is positive.
+    Positive(Value),
+    /// Balance is negative, such transaction can't be valid.
+    Negative(Value),
+    /// Balance is zero.
+    Zero,
+}
+
+fn balance(tx: &tx::Transaction<Address>, fee: Value) -> Result<Balance, ValueError> {
+    let inputs = Value::sum(tx.inputs.iter().map(|i| i.value))?;
+    let outputs = Value::sum(tx.outputs.iter().map(|o| o.value))?;
+    let z = (outputs + fee)?;
+    if inputs > z {
+        Ok(Balance::Positive((inputs - z)?))
+    } else if inputs < z {
+        Ok(Balance::Negative((z - inputs)?))
+    } else {
+        Ok(Balance::Zero)
+    }
+}
+
+pub struct TransactionFinalizer(tx::SignedTransaction<Address>);
+
+impl TransactionFinalizer {
+    fn new(transaction: tx::Transaction<Address>) -> Self {
+        TransactionFinalizer(tx::SignedTransaction {
+            transaction,
+            witnesses: vec![],
+        })
+    }
+
+    /// Sign transaction.
+    pub fn sign(&mut self, pk: &SpendingSecretKey) {
+        // TODO: check if signature is required.
+        // TODO: check if signature matches address.
```

Comment:
this will be proven difficult here as you don't have enough information to check this. The input does not contain the address (so you don't have the public key associated here). Might be worth thinking of other options for future improvements, something like:

```rust
pub fn sign<F>(&mut self, sign: F)
where F: FnMut(&Input) -> Witness
{
    // ...
}
```